### PR TITLE
Issue #1208 - DNSSEC additional items

### LINKED ISSUE #1208: DNSSEC additional items
### Issue description

These are two quick fixes that can be added to the DNSSEC flow to help user experience:

1. Change button copy on modal pop up when user clicks to delete all DS Data. Change button from saying 'Delete all DS Data' to '_Remove_ all DS Data'
2. Change error message for validation on Key tag field. Add validation to Digest field to only accept valid characters. Adjust error messages accordingly for digest.

### Acceptance criteria

- [ ] copy on modal button is changed
- [ ] Validation error message on key tag field reads: "Key tag must be less than 65535"
- [ ] Front end validation on Digest field only accepts alphanumeric characters [(see Slack conversation)](https://cisa-corp.slack.com/archives/C05BDEA3C11/p1699893743536579)

### Additional context

_No response_

### Links to other issues

_No response_

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
## Ticket

Resolves #1208

## Changes

- Copy on modal button changed to 'Remove all DS Data'
- Error message for invalid key tag on DS Data updated to 'Key tag must be less than 65535'
- Digest validation with the following rules:
    - Only hexadecimal characters allowed in digest [0-9, a-f, A-F]
    - Digest length maximum is 64 characters
    - SHA-1 digest must be 40 characters
    - SHA-2 digest must be 64 characters

## Context for reviewers

Test in [BL Sandbox](https://getgov-bl.app.cloud.gov/)

For **design review**

All changes are in the DS Data form under DNSSEC in domain management.

Manage a domain.  From domain management, click DNSSEC, and enable DNSSEC.  This brings up DS Data form.

To test the validation described above:
On the DS Data Form, enter invalid data that doesn't satisfy the conditions described above, and review the error messages.  Error messages should be relevant to the specific form field, and display on the form field and at the top of the form.  The following conditions can be tested:
- empty or unselected values for required fields (key tag, algorithm, digest type, and digest)
- invalid key tag (number greater than 65535; ex.: 65536)
- invalid characters in digest (ex: 0000GG)
- invalid digest length (> 64 characters, ex: 1234567890123456789012345678901234567890123456789012345678901234567890)
- invalid SHA-1 digest ( length != 40; ex: 12345678901234567890123456789012345678)
- invalid SHA-256 digest (length != 256; ex: 12345678901234567890123456789012345678)

Form should also submit successfully when data is valid.  Below are two sample sets of valid ds data:
Key tag: 1234
Algorithm: 3
Digest type: 1 (SHA-1)
Digest: EC0BDD990B39FEEAD889F0BA613DB4ADECB4ADEC

Key tag: 2345
Algorithm: 3
Digest type: 2 (SHA-256)
Digest: EC0BDD990B39FEEAD889F0BA613DB4ADECB4ADECEC0BDD990B39FEFEEAD889F0

Once you have successfully saved one or more ds data records, need to also verify that the modal displays the correct button text.  Using the delete link, delete ds data records until there are no more, then click Save.  A modal should appear.  Verify that the text on the red button reads 'Remove all DS Data'

For **developer code review**

In forms/domain.py - form level validation is in clean; this validates digest based on value in both digest and digest type; field level validation also added for digest; error messages updated for digest type and key tag

utility/errors.py - DsDataError, DsDataErrorCodes, SecurityEmailError, SecurityEmailErrorCodes were added in Issue 1235, which has not yet been merged, but were required for this Issue.  These do not need to be reviewed as part of this PR.

views/domain.py - minor change to button text on modal

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [x] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/forms/domain.py [hunk 1]
```diff
@@ -228,12 +232,16 @@ class DomainDnssecForm(forms.Form):
...
+
     key_tag = forms.IntegerField(
         required=True,
         label="Key tag",
         validators=[
-            MinValueValidator(0, message="Value must be between 0 and 65535"),
-            MaxValueValidator(65535, message="Value must be between 0 and 65535"),
+            MinValueValidator(0, message=str(DsDataError(code=DsDataErrorCodes.INVALID_KEYTAG_SIZE))),
+            MaxValueValidator(65535, message=str(DsDataError(code=DsDataErrorCodes.INVALID_KEYTAG_SIZE))),
         ],
```
Comment (zandercymatics) [inline src/registrar/forms/domain.py:245]: Nice refactor

File: src/registrar/forms/domain.py [hunk 2]
```diff
@@ -228,12 +232,16 @@ class DomainDnssecForm(forms.Form):
 class DomainDsdataForm(forms.Form):
     """Form for adding or editing DNSSEC DS Data to a domain."""
 
+    def validate_hexadecimal(value):
+        if not re.match(r"^[0-9a-fA-F]+$", value):
```
Comment (zandercymatics) [inline src/registrar/forms/domain.py:236]: (Optional) The function name is generally pretty descriptive, but a comment here or function description would be a nice to have that this regex specifically matches a string that consists entirely of one or more hexadecimal digits.

File: src/registrar/forms/domain.py [hunk 3]
```diff
@@ -251,15 +259,40 @@ class DomainDsdataForm(forms.Form):
...
+        cleaned_data = super().clean()
+        digest_type = cleaned_data.get("digest_type", 0)
+        digest = cleaned_data.get("digest", "")
+        # validate length of digest depending on digest_type
+        if digest_type == 1 and len(digest) != 40:
+            self.add_error(
+                "digest",
+                DsDataError(code=DsDataErrorCodes.INVALID_DIGEST_SHA1),
+            )
+        elif digest_type == 2 and len(digest) != 64:
```
Comment (zandercymatics) [inline src/registrar/forms/domain.py:289]: (Q) Given this is essentially a mapping between the type and the expected length, do we expect to add additional digest types here? I am admittedly not too familiar with the internals of DNSSEC.

Comment (dave-kennedy-ecs) [reply src/registrar/forms/domain.py:289]: We only anticipate having these two options.

File: src/registrar/forms/domain.py (hunk 2 above)
Comment (dave-kennedy-ecs) [reply src/registrar/forms/domain.py:242]: added a comment

File: src/registrar/forms/domain.py [hunk 4]
```diff
@@ -228,12 +232,22 @@ class DomainDnssecForm(forms.Form):
 class DomainDsdataForm(forms.Form):
     """Form for adding or editing DNSSEC DS Data to a domain."""
 
+    def validate_hexadecimal(value):
+        """
+        Tests that string matches all hexadecimal values.
+
+        Raise validation error to display error in form
+        if invalid caracters entered
```
Comment (zandercymatics) [inline src/registrar/forms/domain.py:240]:
```suggestion
        if invalid characters entered
```

**Review by zandercymatics — APPROVED** [review]
LGTM but I did find a bug that I believe is unrelated. As I did your code review earlier, you are good on that front minus a minor typo in your comment (noted above). Given I am not as knowledgeable in this area, I would encourage another pair of dev eyes on the code. Specifically just around the "magic numbers" (65535) as, while they seem right (and consistent with what I've looked into), I don't have much experience with DNSSEC in general.

Not sure if its a bug or not, but the "key tag" field will self clear if invalid data is entered inside it. That said I don't think this is related to your PR.
image
image

**Review by therealslimhsiehdy — APPROVED** [review]
LGTM. I wasn't able to reproduce the error Zander posted (I can't type anything other than numbers in the Key Tag area)

Screenshot 2023-11-17 at 1 51 36 PM

**Review by Katherine-Osos — APPROVED** [review]

**Pull request merged**